### PR TITLE
Kiba 2024 Migrator

### PR DESCRIPTION
**Pull request by awisema**

### CODE REVIEW

File: app/models/grda_warehouse/hud/ce_participation.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    belongs_to :data_source, optional: true
+
+    # hide previous declaration of :importable, we'll use this one
+    replace_scope :importable, -> do
+      where(synthetic: false)
+    end
+
+    scope :synthetic, -> do
+      where(synthetic: true)
+    end
```
Comment (eanders) [inline app/models/grda_warehouse/hud/ce_participation.rb]: This table doesn't have a `synthetic` column, pretty sure it shouldn't have these scopes. ^^^

File: app/models/grda_warehouse/hud/hmis_participation.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    belongs_to :data_source, optional: true
+
+    # hide previous declaration of :importable, we'll use this one
+    replace_scope :importable, -> do
+      where(synthetic: false)
+    end
+
+    scope :synthetic, -> do
+      where(synthetic: true)
+    end
```
Comment (eanders) [inline app/models/grda_warehouse/hud/hmis_participation.rb]: synthetic isn't a thing in this table.
```suggestion
    # hide previous declaration of :importable, we'll use this one
    replace_scope :importable, -> do
      where(synthetic: false)
    end

    scope :synthetic, -> do
      where(synthetic: true)
    end
```

File: db/health_structure.sql
Comment (eanders) [inline db/health_structure.sql:1]: Pretty sure this file shouldn't have changed.

File: db/reporting_structure.sql
Comment (eanders) [inline db/reporting_structure.sql:1]: This file shouldn't be changed either.

File: db/structure.sql
Comment (eanders) [inline db/structure.sql:1]: Also, no change here right?

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/concerns/hud_twenty_twenty_two_to_twenty_twenty_four/references.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudTwentyTwentyTwoToTwentyTwentyFour
+  module References
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/concerns/hud_twenty_twenty_two_to_twenty_twenty_four/references.rb:8]: Should this be something like CsvReferences?  I think you mentioned you'd need DB references as well, but this one reads a CSV file, which feels like it's probably not reusable for the db.

Comment (eanders) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/concerns/hud_twenty_twenty_two_to_twenty_twenty_four/references.rb:8]: Oh, maybe that's the model vs file below.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+      row
+    end
+
+    private def entry_enrollment_coc(household_id)
+      @entry_enrollment_coc ||= {}.tap do |h|
+        reference(:enrollment_coc) do |row|
+          next unless row['DataCollectionStage'].to_i == 1
+
+          h[row['HouseholdID']] ||= row['CoCCode']
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb]: I think we'll need to handle empty HouseholdID, if we don't, everyone with a nil HouseholdID will end up with the same CoCCode.  Not sure if we should use EnrollmentID as a backup, or just return a blank CoCCode.  Open to suggestions.

Comment (awisema) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb]: Good point. This seems like a recipe for making bad DQ even worse...
The 2022 spec says:

> There must be one and only one record in EnrollmentCoC.csv with a DataCollectionStage of 1 for each EnrollmentID in Enrollment.csv where RelationshipToHoH is equal to 1. Additionally, there must be one and only one record in EnrollmentCoC.csv with a DataCollectionStage of 1 for each HouseholdID in Enrollment.csv.

So, I don't think using the EnrollmentID will help much. and I don't think leaving it blank would be making the DQ any worse than it already is.

Alternatively, the 2024 spec for the CoCCode column says:

> Must match a record in ProjectCoC.csv with the same ProjectID

So, we could use the code from ProjectCoC when there is only a single CoC Code for the Project given that_most_ projects have only one, if you think it's worth it.

File: db/reporting_structure.sql
Comment (eanders) [inline db/reporting_structure.sql:1]: This one's still here.

File: db/structure.sql
Comment (eanders) [inline db/structure.sql:1]: This one too.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        end
+      end
+      @entry_enrollment_coc[household_id]
+    end
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb]:
```suggestion
    private def entry_enrollment_coc(enrollment_row)
      @entry_enrollment_coc ||= {}.tap do |h|
        reference(:enrollment_coc) do |row|
          next unless row['DataCollectionStage'].to_i == 1

          key = row['HouseholdID'].presence || "en_#{row['EnrollmentID']}" 
          key += "_ds_#{row['data_source_id']}"
          h[key] ||= row['CoCCode'] if HudUtility2024.valid_coc?(row['CoCCode'])
        end
      end
      key = enrollment_row['HouseholdID'].presence || "en_#{enrollment_row['EnrollmentID']}" 
      key += "_ds_#{row['data_source_id']}"
      @entry_enrollment_coc[key]
    end
```

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::AggregatedEnrollment
+  class MoveEnrollmentCoC
+    include ::HudTwentyTwentyTwoToTwentyTwentyFour::References
+
+    def process(row)
+      row['EnrollmentCoC'] = entry_enrollment_coc(row['HouseholdID'])
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/move_enrollment_coc.rb]:
```suggestion
      row['EnrollmentCoC'] = entry_enrollment_coc(row)
```

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/client/update_race_and_ethnicity.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::Client
+  class UpdateRaceAndEthnicity
+    def process(row)
+      row['HispanicLatinaeo'] = row['Ethnicity'] == 1 ? 1 : 0 # 8/9/99 all convert to 0
+      # row['RaceNone'] = nil if row['HispanicLatinaeo'] == 1 # Force clearing of RaceNone if we set HispanicLatinaeo
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/client/update_race_and_ethnicity.rb]: Was this intentionally left commented?

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/csv_transformer.rb [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+          File.copy_stream(file, tmp_file, copy_length, 0)
+          tmp_file.write("\n")
+          tmp_file.close
+        end
+        FileUtils.cp(tmp_file, filename)
+      end
+    ensure
+      tmp_file&.close
+      tmp_file&.unlink
+    end
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/csv_transformer.rb]: What do you think about moving this logic up in to `/app/models/something` or maybe as a monkey patch for CSV?  I feel like we call this pretty often, might be nice to centralize it.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      9 => 9,
+      99 => 99,
+
+      28 => 435,
+      19 => 435,
+      31 => 435,
+      34 => 435,
+      20 => 435,
+      3 => 435,
+    }.freeze
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb]: These appear to be duplicative of `drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/aggregated_enrollment/change_living_situation.rb` Can we centralize that? Maybe in the structure file?

Comment (eanders) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb]: Actually in HudUtility2024 probably makes more sense.  Something like `def cls_migration_for_2022_to_2024`

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/add_rental_subsidy_type.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      19 => 419,
+      31 => 431,
+      34 => 433,
+      20 => 434,
+      3 => 420,
+    }.freeze
+    def process(row)
+      row['RentalSubsidyType'] = nil
+
+      situation = row['LivingSituation'].to_i
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/add_rental_subsidy_type.rb:14]: is `LivingSituation` the right thing to check here?

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/change_living_situation.rb [hunk 11]
```diff
@@ -0,0 +1,64 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::Enrollment
+  class ChangeLivingSituation
+    LIVING_SITUATIONS = {
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/change_living_situation.rb]: Another instance we should DRY

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/move_enrollment_coc.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    private def entry_enrollment_coc(household_id)
+      @entry_enrollment_coc ||= {}.tap do |h|
+        reference(:enrollment_coc) do |row|
+          next unless row['DataCollectionStage'].to_i == 1
+
+          h[row['HouseholdID']] ||= row['CoCCode']
+        end
+      end
+      @entry_enrollment_coc[household_id]
+    end
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/move_enrollment_coc.rb:26]: See the other proposed fix.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/exit/add_destination_subsidy_type.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+module HudTwentyTwentyTwoToTwentyTwentyFour::Exit
+  class AddDestinationSubsidyType
+    SUBSIDY_TYPES = {
+      28 => 428,
+      19 => 419,
+      31 => 431,
+      34 => 433,
+      20 => 434,
+      3 => 420,
+    }.freeze
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/exit/add_destination_subsidy_type.rb]: DRY?

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/exit/update_destination.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HudTwentyTwentyTwoToTwentyTwentyFour::Exit
+  class UpdateDestination
+    DESTINATIONS = {
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/exit/update_destination.rb]: These are probably the same as the living situations.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/hmis_participation/create_hmis_participation.rb [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        yield(row)
+      end
+
+      nil
+    end
+
+    private def victim_service_providers
+      @victim_service_providers ||= {}.tap do |h|
+        reference(:organization) do |row|
+          h[row['OrganizationID']] = row['VictimServiceProvider']
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/hmis_participation/create_hmis_participation.rb]: Need `data_source_id` for the DB version

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/project/update_project_type.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    def process(row)
+      project_type = if row['ProjectType'].to_i == 1
+        if row['TrackingMethod'].to_i == 3
+          1
+        else
+          0
+        end
+      else
+        row['ProjectType']
+      end
```
Comment (eanders) [inline drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/project/update_project_type.rb:18]: Nothing wrong with what you have, but maybe:
```suggestion
      project_type = if row['ProjectType'].to_i == 1 && row['TrackingMethod'].to_i == 3
          0
      else
        row['ProjectType']
      end
```

**Review by eanders — CHANGES_REQUESTED** [review]
A few changes in-line.  Also, maybe we could add some really simple tests where we run a 2022 CSV and convert it to a 2024 and confirm the results match?   Even if we generate the results using the migrator, it would confirm it wasn't throwing an error, and we could eyeball the result file manually to confirm things are processed as expected.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/client/update_race_and_ethnicity.rb (hunk 7 above)
Comment (awisema) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/client/update_race_and_ethnicity.rb]: Oops, no! This was from when I was avoiding stepping on my DB for testing.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb (hunk 9 above)
Comment (awisema) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb]: I moved it to a concern in the migrator -- We shouldn't need it outside of migration, so HUDUtility seems like the wrong place to me.

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/project/update_project_type.rb (hunk 16 above)
Comment (awisema) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/project/update_project_type.rb:18]: I'm going to stick w/ mine... I think yours needed to be `row['TrackingMethod'].to_i == 0` which is fine, but if we both had to work through it to know if it is correct...

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb (hunk 9 above)
Comment (eanders) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/current_living_situation/change_living_situation.rb]: Thanks

File: .github/workflows/rails_tests.yml [hunk 17]
```diff
@@ -193,6 +193,17 @@ jobs:
           aws_region: us-east-1
           flags: --recursive
 
+      - name: 'Fetch 2022 -> 2024 source files'
+        uses: keithweaver/aws-s3-github-action@v1.0.0
+        with:
+          command: cp
+          source: ${{ secrets.AWS_S3_ACTION_BUCKET }}/${{ vars.AWS_S3_ACTIONS_SOURCE_PATH_4_2023 }}
```
Comment (eanders) [inline .github/workflows/rails_tests.yml]: I'm guessing you need 
```suggestion
          source: ${{ secrets.AWS_S3_ACTION_BUCKET }}${{ vars.AWS_S3_ACTIONS_SOURCE_PATH_4_2023 }}
```

File: drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/add_rental_subsidy_type.rb (hunk 10 above)
Comment (awisema) [reply drivers/hud_twenty_twenty_two_to_twenty_twenty_four/app/models/hud_twenty_twenty_two_to_twenty_twenty_four/enrollment/add_rental_subsidy_type.rb:14]: Yes -- the mapping document is looking for 3.917A.1 or 3.917B.1 both of which are stored in `LivingSituation` on Enrollment.

**Review by eanders — APPROVED** [review]

**Review by gigxz — APPROVED** [review]

**Comment by gigxz** [comment]
Looks like there are some test failures in `hud_twenty_twenty_two_to_twenty_twenty_four` fyi

**Comment by gigxz** [comment]
@awisema tests still slightly unhappy 🧐 
would you mind also merging in the latest 86, so I can get a reasonable diff on https://github.com/greenriver/hmis-warehouse/pull/3383? I'm waiting to request review since it's difficult to review in the current state

**Pull request merged**